File: src/main/java/com/koteldlc/client/module/modules/visual/BaseEspModule.java
package com.koteldlc.client.module.modules.visual;

import com.koteldlc.client.gui.settings.BooleanSetting;
import com.koteldlc.client.gui.settings.ColorSetting;
import com.koteldlc.client.gui.settings.SliderSetting;
import com.koteldlc.client.module.Category;
import com.koteldlc.client.module.Module;
import com.mojang.blaze3d.systems.RenderSystem;

/**
 * База для ESP-модулей. Содержит общие настройки цвета/толщины/мигания.
 */
public abstract class BaseEspModule extends Module {
    protected final ColorSetting color;
    protected final SliderSetting thickness;
    protected final BooleanSetting blink;

    protected BaseEspModule(String name, String description, int defaultColor) {
        super(name, description, -1, Category.RENDER);
        this.color = register(new ColorSetting("color", defaultColor));
        this.thickness = register(new SliderSetting("thickness", 2.0, 1.0, 6.0));
        this.blink = register(new BooleanSetting("blink", false));
    }

    public void prepareRenderState() {
        RenderSystem.enableBlend();
        RenderSystem.defaultBlendFunc();
        RenderSystem.lineWidth((float) thickness.getValue().doubleValue());
    }
}

File: kotlin-mod/src/main/kotlin/com/koteldlc/client/modules/visual/BaseEspModule.kt
package com.koteldlc.client.modules.visual

import com.koteldlc.client.modules.Category
import com.koteldlc.client.modules.Module
import com.mojang.blaze3d.systems.RenderSystem

abstract class BaseEspModule(name: String) : Module(name, Category.VISUAL) {
    init {
        settings["color"] = 0xFFFF4444.toInt()
        settings["thickness"] = 2.0
        settings["blink"] = false
    }

    fun renderEsp() {
        if (!enabled) return
        RenderSystem.enableBlend()
        RenderSystem.defaultBlendFunc()
        RenderSystem.lineWidth((settings["thickness"] as Double).toFloat())

        val argb = settings["color"] as Int
        val a = ((argb shr 24) and 0xFF) / 255.0f
        val r = ((argb shr 16) and 0xFF) / 255.0f
        val g = ((argb shr 8) and 0xFF) / 255.0f
        val b = (argb and 0xFF) / 255.0f
        RenderSystem.setShaderColor(r, g, b, a)

        if (settings["blink"] as Boolean) {
            // Здесь можно добавить пульсацию альфа-канала/яркости.
        }

        onRenderEsp()
        RenderSystem.disableBlend()
    }

    protected abstract fun onRenderEsp()
}
